### ISSUE TITLE
Autosave idea

### ISSUE THREAD

**Issue by NoelCov** [freeCodeCamp/freeCodeCamp#50633]
### Is your feature request related to a problem? Please describe.

II recently heard from a friend who is doing one of the **Responsive Web Design certification projects** that their browser shut down and when they re-opened it the code that they had been working on was lost. I am sure that this probably happens to people often since everyone runs into problems with their browsers and or computers.

### Describe the solution you'd like

I think a good solution would be to implement auto save.

### Describe alternatives you've considered

I have considered that instead of auto saving to the database, maybe we could keep it in the browser's local storage for when the browser is closed for whatever reason at least we can retrieve the code from there and only save when the user clicks the "Save your code" button. This would make it a bit easier I think since we will not be writing to the database too often for every user.

### Additional context

For context this is my first time opening a new issue, it is something that I thought would be cool to add and I also would love to work on it if given the chance. I wanted to first post the issue though and see if it's something that we can actually work on.

**Comment by elliezub** [comment]
Great idea to have an auto-save feature! This happened to me when I was doing the course a while ago. I think saving to local storage is good, but if you are just learning about HTML and CSS for the first time, I think you would not know how to access that. So people would most likely just start over in that case.

**Comment by raisedadead** [comment]
There have been several discussions about this over the years (I would recommend a search if you are interested in the conversations), and largely it boils down to these limitations:

1. Browser-based storage is too brittle and unreliable.
2. We do not have the infra budget to store anything other than the code that counts for the "certifications" on a DB.
3. The UX around what is auto-saved and auto-loaded (or on-demand) is complex and needs engineering time, which we (as staff) are limited on.
4. The newer project-based curricular content doesn't need to save code. This is by design for pedagogical reasons.
5. Some older certifications already have some of these auto-saving features.

That said, we are open to the idea as always (over the years), but it needs to check a few boxes that include but are not limited to:

- [ ] Simple UX that is clear to the users.
- [ ] The practice projects SHOULD NOT save code BY DESIGN (as in pedagogic reasons).
- [ ] The remote DB (server) can't be used for cost reasons.
- [ ] Using local storage is fine as long as it is more reliable, as users sign in and out, clean up caching, and a dozen other things you can look up in past discussions.

Finally, if you are interested in building this, we welcome a PR, but please consider that this will add additional maintenance overhead. The staff believes that the freeCodeCamp's editor is not meant to offer IDE-like features - but not against ideas like these.

If you want more help, don't hesitate to contact us in the contributor's chat!

**Comment by raisedadead** [comment]
Yup - I personally would rely on something other than local storage. As for the nature of the project code, if the code is used for evaluation and awarding a certification – then yes, those should be saved to the DB (all the time, including for the intermediate steps).

**Comment by moT01** [comment]
I think the terms "Practice projects" and "Certification projects" are too vague for this issue. There are many different types of each of those.

I think it might be useful to talk about this in terms of save methods and saved content rather than challenge types. Many different challenge types use the same save method. Here's a breakdown of the different challenges and methods that are relevant here. They only include challenges that use our editor. For all other challenges, a URL is submitted, or just a record of completion is saved, etc. (so there's no code to save).

| | Challenges | Description/Examples | Cert Project? | On Save | On submit |
|-|------------|-----------------------|--------------|----------|-----------|
| 1 | RWD Practice Challenges & Classic Challenges | [Steps for the RWD practice projects](https://www.freecodecamp.org/learn/2022/responsive-web-design/learn-html-by-building-a-cat-photo-app/step-1), [most of the challenges in the "front 4"](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/basic-javascript/comment-your-javascript-code), [project Euler](https://www.freecodecamp.org/learn/project-euler/project-euler-problems-1-to-100/problem-1-multiples-of-3-and-5), etc. | No | Code saved to local storage | Code not saved to DB |
| 2 | Javascript Certification Projects | Same as the "Classic Challenges" above, but they save campers solution to the DB - [there's only five of these](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/javascript-algorithms-and-data-structures-projects/palindrome-checker) | Yes | Code saved to local storage | Code saved to DB, `completedChallenges[]` |
| 3 | RWD Certification Projects | Uses our multi-file editor - [there's only five of these](https://www.freecodecamp.org/learn/2022/responsive-web-design/build-a-survey-form-project/build-a-survey-form) |  Yes | Code saved to DB, `savedChallenges[]` | Code saved to DB, `completedChallenges[]` |

Hopefully that doesn't just make things more confusing. The only type the OP and I were talking about (correct me if I'm wrong @NoelCov), are the RWD Certification Projects. The issue is that campers can write a lot of code for these and easily lose it if they don't manually save their code (`ctrl/cmd+s` or click `save`). i.e. browser/computer crash, accidental browser back button click, accidental refresh, navigating through UI buttons, etc. If any of these happen, the code is gone. Hopefully, that defines the problem better.

Some potential solutions:
- Auto-save to database
- Auto-save to local storage
- Give campers a message if they haven't saved in a while
- Attempt to intercept campers from accidentally leaving (browser back, other UI buttons, etc.)

Perhaps one or a combination of these could help with the problem.

Type 1 challenges in the table above don't need any changes. We aren't going to save the code for those to the DB any time soon.
Type 2 in the table above could potentially use some similar improvements - but I think these will turn into type 3 when the new JS cert is finished.

**Comment by moT01** [comment]
The problem with auto-saving is that if I write some code, save it, write some more code and I don't like that new code so I want to go back to my last save - I wouldn't be able to do that.

So I think the best solution here is to try and stop campers from leaving the page if they have unsaved changes. That might be tricky. I have a few methods in the exam environment to stop the browser back and browser close, but that doesn't have the nav, so it doesn't have to handle gatsby buttons - not sure how easy it will be to handle those, but I think we should make an attempt at going that route.

**Comment by X1Vi** [comment]
How about we allow users to save their code locally on their machine with a download button or something ?
So they can just copy and paste their code later on.

We can save only the current code and not any other previous code just in case if a user wishes to recover it if   the browser gets closed.

**Comment by srinivas-batthula** [comment]
Hi, Is this issue still open? if yes, i would like to take it up, please assign it to me.

**Comment by ilenia-magoni** [comment]
hi @srinivas-batthula We typically do not assign issues. Instead, we accept the first pull request that comprehensively solves the issue.

Issues labeled with `help wanted` or `first timers only` are open for contributions.

Please make sure you read [our guidelines for contributing](https://contribute.freecodecamp.org/#/). We prioritize contributors following the instructions in our guide. Join us in [our chat room](https://discord.gg/PRyKn3Vbay) or [the forum](https://forum.freecodecamp.org/c/contributors/3) if you need help contributing - our community will be happy to assist you.

**Comment by anish-eng** [comment]
is this issue still open for accepting contributions?

**Comment by huyenltnguyen** [comment]
We have decided not to implement auto-save and will show a warning modal instead.

Closing in favor of #54796.
